File: mawhrin/items/note/note.go
// Package note implements the Note type and methods.
package note

import (
	"os"

	"github.com/stvmln86/mawhrin/mawhrin/tools/file"
	"github.com/stvmln86/mawhrin/mawhrin/tools/neat"
	"github.com/stvmln86/mawhrin/mawhrin/tools/path"
)

// Note is a single plaintext file in a directory.
type Note struct {
	Orig string
	Mode os.FileMode
}

// New returns a new Note.
func New(orig string, mode os.FileMode) *Note {
	return &Note{orig, mode}
}

// Delete moves the Note to the ".trash" extension.
func (n *Note) Delete() error {
	return file.Delete(n.Orig)
}

// Exists returns true if the Note exists.
func (n *Note) Exists() bool {
	return file.Exists(n.Orig)
}

// Match returns true if the Note's name contains a substring.
func (n *Note) Match(term string) bool {
	return path.Match(n.Orig, term)
}

// Name returns the Note's name.
func (n *Note) Name() string {
	return neat.Name(path.Name(n.Orig))
}

// Read returns the Note's body as a string.
func (n *Note) Read() (string, error) {
	body, err := file.Read(n.Orig)
	return neat.Body(body), err
}

// Search returns true if the Note's body contains a string.
func (n *Note) Search(term string) (bool, error) {
	return file.Search(n.Orig, term)
}

// Update overwrites the Note's body with a string.
func (n *Note) Update(body string) error {
	body = neat.Body(body)
	return file.Update(n.Orig, body, n.Mode)
}

File: mawhrin/items/note.py
"""
Class definition for 'Note'.
"""

from collections.abc import Iterator

from mawhrin import tools


class Note:
    """
    A single plaintext note file in a directory.
    """

    def __init__(self, path: str):
        """
        Initialise the Note.
        """

        self.path = path

    def __eq__(self, note: object) -> bool:
        """
        Return True if the Note is equal to another Note.
        """

        if not isinstance(note, Note):
            return NotImplemented
        else:
            return self.path == note.path

    def __hash__(self) -> int:
        """
        Return the Note's unique hashing integer.
        """

        return hash("Note:" + self.path)

    def __iter__(self) -> Iterator[str]:
        """
        Yield each line of the Note's body as strings.
        """

        yield from self.read().splitlines(keepends=True)

    def __repr__(self) -> str:
        """
        Return the Note as a code-representative string.
        """

        return f"Note({self.path!r})"

    def __str__(self) -> str:
        """
        Return the Note as a descriptive string.
        """

        return self.name

    def delete(self):
        """
        Delete the Note by renaming it to the ".trash" extension.
        """

        tools.file.delete(self.path)

    def exists(self) -> bool:
        """
        Return True if the Note's file exists.
        """

        return tools.file.exists(self.path)

    def match(self, text: str) -> bool:
        """
        Return True if the Note's name contains a case-insensitive substring.
        """

        return tools.path.match(self.path, text)

    def read(self) -> str:
        """
        Return the Note's body as a string.
        """

        body = tools.file.read(self.path)
        return tools.data.body(body)

    def rename(self, name: str):
        """
        Rename the Note to a new name with the same directory and extension.
        """

        name = tools.data.name(name)
        tools.file.rename(self.path, name)

    def search(self, text: str) -> bool:
        """
        Return True if the Note's body contains a case-insensitive substring.
        """

        return tools.file.search(self.path, text)

    @property
    def name(self) -> str:
        """
        Return the Note's base name without the extension.
        """

        name = tools.path.name(self.path)
        return tools.data.name(name)

    def write(self, body: str):
        """
        Overwrite the Note's body with a string.
        """

        body = tools.data.body(body)
        tools.file.write(self.path, body)
